man/kasan.1
.TH  Tukubai 1 "28 Dec 2025" "usp Tukubai" "Tukubai コマンド マニュアル"

.br
kasan(1)
.br

.br
【名前】
.br

.br
kasan\ :\ 累計
.br

.br
【書式】
.br

.br
Usage\ \ \ :\ kasan\ [+r]\ [key=<key>]\ val=<n>\ [<file>]
.br
Version\ :\ Sun\ Dec\ 28\ 02:56:32\ JST\G 2025
.br

.br
【説明】
.br

.br
引数のファイルまたは標準入力のテキストデータの指定したフィールドの値を、各行ご
.br
とに加算して指定したフィールドの次のフィールドに挿入する。累計値を求める場合な
.br
どに使用する。
.br
<file>として-を指定するか省略すると標準入力を使用する。
.br

.br
【例1】
.br

.br
店別の累計の仕入れ値を算出する。ここでは4フィールド目の値を加算して5フィールド
.br
目に挿入する。
.br

.br

  $ cat data
  a店 1日目 103 62                ←店名 日付 売れ数 仕入数
  a店 2日目 157 94
  a店 3日目 62 30
  a店 4日目 131 84
  a店 5日目 189 111
  a店 6日目 350 20
  a店 7日目 412 301

  $

.br

  $ kasan key=4 data
  a店 1日目 103 62 62
  a店 2日目 157 94 156
  a店 3日目 62 30 186
  a店 4日目 131 84 270
  a店 5日目 189 111 381
  a店 6日目 350 20 401
  a店 7日目 412 301 702

  $

.br
【例2】
.br

.br
あるフィールドの値が変わった場所で加算をリセットする場合は、そのリセットのキー
.br
となるフィールドをkey=で指定する。たとえば店別の累計仕入数を算出する場合には、
.br
次のように1フィールド目が変更されるごとに加算をリセットする。
.br

.br
key=<key>で指定できるのは数値、範囲(/)、NF\ NF-<n>である。
.br

.br

  $ cat data
  a店 1日目 103 62                ←店名 日付 売れ数 仕入数
  a店 2日目 157 94
  a店 3日目 62 30
  a店 4日目 131 84
  b店 1日目 210 113
  b店 2日目 237 121
  b店 3日目 150 82
  b店 4日目 198 105
  c店 1日目 81 52
  c店 2日目 76 49
  c店 3日目 38 21
  c店 4日目 81 48

  $

.br

  $ kasan key=1 val=4 data
  a店 1日目 103 62 62
  a店 2日目 157 94 156
  a店 3日目 62 30 186
  a店 4日目 131 84 270
  b店 1日目 210 113 113           ←累計加算をリセット
  b店 2日目 237 121 234
  b店 3日目 150 82 316
  b店 4日目 198 105 421
  c店 1日目 81 52 52              ←累計加算をリセット
  c店 2日目 76 49 101
  c店 3日目 38 21 122
  c店 4日目 81 48 17

  $

.br
【例3】
.br

.br
元データのフィールドをその加算値と置き換えて出力する場合は+rオプションを使用す
.br
る。たとえば仕入れ数を累計仕入れ数に置き換えるのであれば、この場合であれば4フィ
.br
ールド目の加算値を4フィールド目に出力するようにすれば良い。
.br

.br

  $ cat data              ←店名 日付 売れ数 仕入数
  a店 1日目 103 62
  a店 2日目 157 94
  a店 3日目 62 30
  a店 4日目 131 84
  a店 5日目 189 111
  a店 6日目 350 20
  a店 7日目 412 301

  $

.br

  $ kasan +r key=4 data
  a店 1日目 103 62
  a店 2日目 157 156
  a店 3日目 62 186
  a店 4日目 131 270
  a店 5日目 189 381
  a店 6日目 350 401
  a店 7日目 412 702

  $

.br
【例4】
.br

.br
小数点のあるデータを加算する場合、それまでの行にある最大小数桁数に合わせて出力
.br
するようになる。
.br

.br

  $ cat data
  A 1
  A 1.2
  A 1.23
  B 2
  B 2.34
  B 3

  $

.br

  $ kasan key=1 val=2 data
  A 1 1
  A 1.2 2.2
  A 1.23 3.43
  B 2 2
  B 2.34 4.34
  B 3 7.34

  $

.br
【例5】
.br

.br
key=<pos>で指定できるのは数値、範囲(/)、列挙(@)、NF\ NF-<n>である。
.br

.br

  $ kasan key=1 data

.br
第2から第3フィールドの各フィールドを加算し、加算値をそれぞれのフィールドの直後
.br
に挿入するには次のようにコマンドを実行する。
.br

.br

  $ cat data
  A 1 2
  A 3 4
  A 5 6

  $

.br

  $ kasan key=2/3 data
  A 1 1 2 2
  A 3 4 4 6
  A 5 9 6 12

  $

.br
第2フィールドと第4フィールドの各フィールドを加算し、加算値をそれぞれのフィール
.br
ドの直後に挿入するには次のようにコマンドを実行する。
.br

.br

  $ kasan key=2@4 data

.br
次のようにNFも指定できる。
.br

.br

  $ kasan key=NF data

.br

  $ kasan key=NF-2 data

.br
【関連項目】
.br

.br
plus(1)、ratio(1)、sm2(1)、sm4(1)、sm5(1)、ysum(1)
.br

.br
last\ modified:\ Sun\ Dec\ 28\ 03:09:16\ JST\ 2025
.br
Contact\ us:\ uecinfo@usp-lab.com
.br
Copyright\ (c)\ 2012-2025\ Universal\ Shell\ Programming\ Laboratory\ All\ Rights
.br


COMMANDS.HS/kasan.hs
#!/usr/bin/env runghc
import System.Environment
import System.IO
import Text.ParserCombinators.Parsec
import Control.Monad
import Data.ByteString.Lazy.Char8 as BS hiding (filter,head,drop,take,reverse)
import Control.Applicative hiding ((<|>), many)

{--
kasan（Open usp Tukubai）

designed by Nobuaki Tounaka
written  by Hinata Yanagi

The MIT License

Copyright (C) 2025 Universal Shell Programming Laboratory

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
--}

showUsage :: IO ()
showUsage = do System.IO.hPutStr stderr ("Usage    : kasan [+r] [ref=<ref>] key=<n> [<file>]\n" ++ 
                "Version  : Mon Jan 20 17:18:16 JST 2025\n" ++
                "Open usp Tukubai (LINUX+FREEBSD)\n")

main :: IO ()
main = do args <- getArgs
          case args of
              []            -> showUsage
              ["-h"]        -> showUsage
              ["--help"]    -> showUsage
              ["--version"] -> showUsage
              _          -> mainProc (setOpts args)

mainProc :: Opts -> IO ()
mainProc (Opts rep ref key file) = readF file >>= mainProc' rep ref key


mainProc' :: Bool -> Int -> [Int] -> BS.ByteString -> IO ()
mainProc' rep ref ks cs = out rep [ parseLine ref ks ln | ln <- BS.lines cs ] nullRecord
 where nullRecord = Record [] (BS.pack "") [] []

out :: Bool -> [Record] -> Record -> IO ()
out rep []     _    = do return ()
out rep (r:rs) prep = BS.putStrLn (out' rep newr) >> out rep rs newr
                      where newr = newRecord r prep

out' :: Bool -> Record -> BS.ByteString
out' rep (Record ws refstr fs vs) = BS.unwords $ out'' [1..(Prelude.length ws)] rep fs vs ws

out'' :: [Int] -> Bool -> [Int] -> [Double] -> [BS.ByteString] -> [BS.ByteString]
out'' []     _   _  _  _  = []
out'' (n:ns) rep fs vs ws = ans : out'' ns rep fs rvs ws
        where x = filter (==n) fs
              w = ws !! (n-1)
              nstr = BS.pack $ formatNum $ show $ head vs
              repstr = if rep then nstr else BS.unwords [w,nstr]
              ans = if x == [] then w else repstr
              rvs = if x == [] then vs else (drop 1 vs)

formatNum :: String -> String
formatNum str = if x == [] then str else reverse $ cutZero (reverse str)
        where x = filter (== '.') str

cutZero :: String -> String
cutZero ('.':cs) = cs
cutZero ('0':cs) = cutZero cs
cutZero (c:cs)   = c:cs

newRecord :: Record -> Record -> Record
newRecord (Record ws refstr fs vs) (Record pws prefstr pfs pvs)
  | pvs == []         = Record ws refstr fs vs
  | refstr /= prefstr = Record ws refstr fs vs
  | refstr == prefstr = Record ws refstr fs [ (fst x) + (snd x) | x <- Prelude.zip vs pvs]

parseLine :: Int -> [Int] -> BS.ByteString -> Record
parseLine ref ks ln = Record ws refstr nks vs
          where ws = myWords ln
                nks = [ if k > 0 then k else k + (Prelude.length ws) | k <- ks ]
                refstr = if ref == 0 then (BS.pack "") else (ws !! (ref-1))
                vs = f nks ws
                f [] _      = []
                f (k:nks) ws = (read (BS.unpack $ ws !! (k-1))::Double ) : f nks ws

myWords :: BS.ByteString -> [BS.ByteString]
myWords line = filter (/= x) $ BS.split ' ' line
               where x = BS.pack ""

data Record = Record [BS.ByteString] BS.ByteString [Int] [Double] deriving Show

readF :: String -> IO BS.ByteString
readF "-" = BS.getContents
readF f   = BS.readFile f

data Opts = Opts Bool Int [Int] String | Error String deriving Show

setOpts :: [String] -> Opts
setOpts as = case parse args "" ((Prelude.unwords as) ++ " ") of
                  Right opt -> opt
                  Left err -> Error ( show err )

args = Opts <$> (try(string "+r " >> return True) <|> return False)
            <*> (try(ref) <|> return 0)
            <*> (try(key) <|> return [0])
            <*> (try(filename) <|> return "-")

ref = do string "ref="
         a <- many1 digit
         char ' '
         return (read a)

key = do string "key="
         a <- try(atnums) <|> try(slashnums) <|> simplepos
         char ' '
         return a

atnums = do a <- posnum
            nums <- many1 atnum
            return (a:nums)

atnum = char '@' >> posnum

slashnums = do a <- posnum
               char '/'
               b <- posnum
               return [a..b]

simplepos = do a <- posnum
               return [a]

posnum = try(nfmnum) <|> try(nfnum) <|> intnum

intnum = many1 digit >>= return . read

nfnum = string "NF" >> return 0

nfmnum = do string "NF-"
            a <- many1 digit 
            return $ (read a) * (-1)


filename =  many1 ( letter <|> digit <|> symbol )

symbol = oneOf "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
